pkg/addonmanager/controllers/agentdeploy/hosted_hook_sync.go

```diff
@@ -80,20 +80,31 @@ func (s *hostedHookSyncer) sync(ctx context.Context,
 		return addon, nil
 	}
 
-	// will deploy the pre-delete hook manifestWork when the addon is deleting
 	if addon.DeletionTimestamp.IsZero() {
 		addonAddFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer)
 		return addon, nil
 	}
 
-	// the hook work is completed if there is no HostingPreDeleteHookFinalizer when the addon is deleting.
 	if !addonHasFinalizer(addon, addonapiv1alpha1.AddonHostingPreDeleteHookFinalizer) {
 		return addon, nil
 	}
 
-	hookWork, err = s.applyWork(ctx, addonapiv1alpha1.ManagedClusterAddOnHostingManifestApplied, hookWork, addon)
-	if err != nil {
-		return addon, err
+	// apply the pre-delete hook manifestWork when the addon is deleting and HookManifestCompleted condition is not true.
+	// there are 2 cases: one is the HookManifestCompleted condition is false,another is there is no this condition.
+	if !meta.IsStatusConditionTrue(addon.Status.Conditions, addonapiv1alpha1.ManagedClusterAddOnHookManifestCompleted) {
+		hookWork, err = s.applyWork(ctx, addonapiv1alpha1.ManagedClusterAddOnHostingManifestApplied, hookWork, addon)
+		if err != nil {
+			return addon, err
+		}
+	} else {
+		// cleanup is safe here since there is no case which HookManifestCompleted condition is changed from true to false.
```

Comment:
if it fails to cleanup hook work/remove finalizer here, it also could result in creating the hook again, right?

Reply:
no, if got to clean part, that means the hookcompleted condition is true, so that means the addon after this in the queue has the true hookcompleted condition, that will never go to apply manifestwork part.

Reply:
if it fails at the cleanup part, it returns, so no chance to add the hook completed condition right?
But it is a corner case.

Reply:
no, if it fails at the cleanup part, that means the  hookcompleted condition has been true. no place to change the condition to false.

Reply:
got it!